webdriver-framework-junit4/src/main/java/com/browserstack/webdriver/junit4/WebDriverProviderRule.java

```diff
@@ -0,0 +1,71 @@
+package com.browserstack.webdriver.junit4;
+
+import java.net.MalformedURLException;
+import com.browserstack.webdriver.config.DriverType;
+import org.junit.rules.TestWatcher;
+import org.junit.runner.Description;
+import org.openqa.selenium.JavascriptExecutor;
+import org.openqa.selenium.WebDriver;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import com.browserstack.webdriver.config.Platform;
+import com.browserstack.webdriver.core.WebDriverFactory;
+
+public class WebDriverProviderRule extends TestWatcher {
+    private static final Logger LOGGER = LoggerFactory.getLogger(WebDriverProviderRule.class);
+    private static final String TEST_STATUS_SCRIPT = "browserstack_executor: {\"action\": \"setSessionStatus\", \"arguments\": {\"status\": \"%s\", \"reason\": \"%s\"}}";
+    private String methodName;
+    private WebDriver driver;
+    private final WebDriverFactory webDriverFactory;
+
+    public WebDriverProviderRule() {
+        this.webDriverFactory = WebDriverFactory.getInstance();
+    }
+
+    /**
+     * @return the name of the currently-running test method
+     */
+    public WebDriver getWebDriver(Platform platform) throws MalformedURLException {
+        this.driver = webDriverFactory.createWebDriverForPlatform(platform, this.methodName);
+        return driver;
+    }
+
+    /**
+     * Invoked when a test succeeds
+     */
+    protected void succeeded(Description description) {
+        LOGGER.info("Succeeded Test :: {} WebDriver Session :: {}", description.getDisplayName(), this.driver);
+        if(webDriverFactory.getDriverType() == DriverType.cloudDriver){
+            ((JavascriptExecutor) driver).executeScript(String.format(TEST_STATUS_SCRIPT, "passed", "Test Passed"));
+        }
+    }
+
+    /**
+     * Invoked when a test fails
+     */
+    protected void failed(Throwable e, Description description) {
+        LOGGER.info("Failed Test :: {} WebDriver Session :: {}", description.getDisplayName(), this.driver, e);
+        if(webDriverFactory.getDriverType() == DriverType.cloudDriver){
+            ((JavascriptExecutor) driver).executeScript(String.format(TEST_STATUS_SCRIPT, "failed", e.getMessage()));
+        }
```

Comment:
This does not seem to provide a very relevant error message. Please try it out once for a missing element or some such. You can also look at the implementation in the testng glue code if required.

Reply:
This comes up as the error message.
```no such element: Unable to locate element: {"method":"css selector","selector":"#login\-btn1"}
  (Session info: chrome=92.0.4515.107)
For documentation on this error, please visit: https://www.seleniumhq.org/exceptions/no_such_element.html
Build info: version: '3.141.59', revision: 'e82be7d358', time: '2018-11-14T08:17:03'
System info: host: 'Abhi-Singhs-MacBook-Pro.local', ip: 'fe80:0:0:0:1071:81da:337f:1573%en0', os.name: 'Mac OS X', os.arch: 'x86_64', os.version: '10.16', java.version: '11.0.10'```
Using the first line of the error message as Failure reason.